includes/class-sensei-guest-user.php

```diff
@@ -79,22 +97,93 @@ public function open_course_enable_course_access( $can_view_course_content, $cou
 	 * Create a guest user for open access courses if no user is logged in.
 	 *
 	 * @since $$next-version$$
+	 * @access private
 	 */
 	public function sensei_create_guest_user_and_login_for_open_course() {
 		global $post;
 
-		// Conditionally create Guest Student user and set role for open course.
+		// Conditionally creates or retrieves Guest Student user and set role for open course.
 		if (
 			$this->is_take_course_action()
 			&& ! is_user_logged_in()
 			&& $this->is_course_open_access( $post->ID )
 		) {
-			$user_id = $this->create_guest_student_user();
+			$user_id = $this->create_new_or_retrieve_existing_guest_student();
 			$this->login_user( $user_id );
 			$this->recreate_nonces();
 		}
 	}
 
+	/**
+	 * Logs guest user out if out of open course context.
+	 *
+	 * @since $$next-version$$
+	 * @access private
+	 */
+	public function sensei_log_guest_user_out_if_not_open_course_related_action() {
+		if (
+			is_user_logged_in() &&
+			$this->is_current_user_guest() &&
+			! $this->is_open_course_related_action()
+		) {
+			wp_logout();
+		}
+	}
+
+	/**
+	 * Automatically logs guest user in if user already enrolled in that course as guest user.
+	 *
+	 * @since $$next-version$$
+	 * @access private
+	 */
+	public function sensei_auto_log_guest_in_if_already_enrolled() {
+		if (
+			! is_user_logged_in() &&
+			$this->is_open_course_related_action() &&
+			$this->is_guest_student_exists() &&
+			Sensei_Course::is_user_enrolled( $this->get_course_id_for_course_related_pages(), $this->get_guest_student_id() )
+		) {
+			$this->login_user( $this->get_guest_student_id() );
+			$this->recreate_nonces();
+		}
+	}
+
+	/**
+	 * Checks if the action is related to an open course or a lesson or a quiz that belongs to an open course.
+	 *
+	 * @since $$next-version$$
+	 * @access private
+	 * @return boolean
+	 */
+	private function is_open_course_related_action() {
+		if ( ! is_singular( [ 'course', 'lesson', 'quiz' ] ) ) {
+			return false;
+		}
+
+		return $this->is_course_open_access( $this->get_course_id_for_course_related_pages() );
+	}
+
+	/**
+	 * Finds out the course id for open course, lesson and quiz pages.
+	 *
+	 * @since $$next-version$$
+	 * @access private
+	 * @return int
+	 */
+	private function get_course_id_for_course_related_pages() {
```

Comment:
There is already a helper for this we can use: [`Sensei_Utils::get_current_course()`](https://github.com/Automattic/sensei/blob/397940d94baa97d13c20268192bd0120da8b6a81/includes/class-sensei-utils.php#L2194)

Reply:
Thanks! Updated here https://github.com/Automattic/sensei/pull/6280/commits/c91700a06c585e0bb5b5960004e04a8208f76f0b